File: src/test/java/com/plog/domain/image/controller/ImageControllerTest.java
package com.plog.domain.image.controller;

import com.plog.domain.image.dto.ImageUploadRes;
import com.plog.domain.image.service.ImageService;
import com.plog.global.security.SecurityUser;
import com.plog.testUtil.SecurityTestConfig;
import com.plog.testUtil.WebMvcTestSupport;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest;
import org.springframework.context.annotation.Import;
import org.springframework.http.MediaType;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.ResultActions;

import java.util.List;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static org.mockito.ArgumentMatchers.anyLong;

/**
 * ImageController의 웹 계층 단위 테스트입니다.
 */
@WebMvcTest(ImageController.class)
@ActiveProfiles("test")
@Import(SecurityTestConfig.class) // [추가] SecurityTestConfig 가져오기
class ImageControllerTest extends WebMvcTestSupport {

    @MockitoBean
    private ImageService imageService;

    // [추가] 테스트 실행 전 가짜 인증 정보 주입
    @BeforeEach
    void setUpUser() {
        // 1. SecurityUser를 Mock으로 생성 (생성자 로직 회피)
        SecurityUser mockUser = mock(SecurityUser.class);

        // 2. 인증 토큰 생성
        UsernamePasswordAuthenticationToken authentication =
                new UsernamePasswordAuthenticationToken(mockUser, null, null);

        // 3. SecurityContext에 설정 (컨트롤러의 @AuthenticationPrincipal이 이 정보를 읽어감)
        SecurityContextHolder.getContext().setAuthentication(authentication);
    }

    @Test
    @DisplayName("이미지 업로드 성공 시 successUrls를 포함한 응답을 반환한다")
    void uploadImageSuccess() throws Exception {
        MockMultipartFile file = new MockMultipartFile(
                "file", "test.jpg", "image/jpeg", "test data".getBytes()
        );

        // ✅ ImageUploadRes Mock
        ImageUploadRes mockResult = new ImageUploadRes(
                List.of("http://minio/bucket/images/uuid.jpg"),
                List.of()
        );
        given(imageService.uploadImage(any(), anyLong())).willReturn(mockResult);

        ResultActions resultActions = mockMvc
                .perform(
                        multipart("/api/images")
                                .file(file)
                                .contentType(MediaType.MULTIPART_FORM_DATA)
                )
                .andDo(print());

        resultActions
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status").value("success"))
                .andExpect(jsonPath("$.data.successUrls[0]").value("http://minio/bucket/images/uuid.jpg"))
                .andExpect(jsonPath("$.data.successUrls").isArray())
                .andExpect(jsonPath("$.data.failedFilenames").isEmpty())
                .andExpect(jsonPath("$.message").value("이미지 업로드 성공"));
    }

    @Test
    @DisplayName("다중 이미지 업로드 성공 시 successUrls 리스트를 반환한다")
    void uploadImagesSuccess() throws Exception {
        MockMultipartFile file1 = new MockMultipartFile(
                "files", "test1.jpg", "image/jpeg", "data1".getBytes()
        );
        MockMultipartFile file2 = new MockMultipartFile(
                "files", "test2.jpg", "image/jpeg", "data2".getBytes()
        );

        // ✅ ImageUploadRes Mock
        ImageUploadRes mockResult = new ImageUploadRes(
                List.of(
                        "http://minio/bucket/images/uuid1.jpg",
                        "http://minio/bucket/images/uuid2.jpg"
                ),
                List.of()
        );
        given(imageService.uploadImages(anyList(), anyLong())).willReturn(mockResult);

        ResultActions resultActions = mockMvc
                .perform(
                        multipart("/api/images/bulk")
                                .file(file1)
                                .file(file2)
                                .contentType(MediaType.MULTIPART_FORM_DATA)
                )
                .andDo(print());

        resultActions
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.status").value("success"))
                .andExpect(jsonPath("$.data.successUrls").isArray())
                .andExpect(jsonPath("$.data.successUrls.length()").value(2))
                .andExpect(jsonPath("$.data.successUrls[0]").value("http://minio/bucket/images/uuid1.jpg"))
                .andExpect(jsonPath("$.data.successUrls[1]").value("http://minio/bucket/images/uuid2.jpg"))
                .andExpect(jsonPath("$.data.failedFilenames").isEmpty())
                .andExpect(jsonPath("$.message").value("다중 이미지 업로드 성공"));
    }

    @Test
    @DisplayName("다중 이미지 업로드 부분 실패 시 실패 파일명도 포함하여 반환한다")
    void uploadImagesPartialFailure() throws Exception {
        MockMultipartFile file1 = new MockMultipartFile("files", "test1.jpg", "image/jpeg", "data1".getBytes());

        ImageUploadRes mockResult = new ImageUploadRes(
                List.of("http://minio/uuid1.jpg"),
                List.of("invalid.txt")
        );
        given(imageService.uploadImages(anyList(), anyLong())).willReturn(mockResult);

        mockMvc.perform(multipart("/api/images/bulk").file(file1))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.successUrls").isArray())
                .andExpect(jsonPath("$.data.successUrls.length()").value(1))
                .andExpect(jsonPath("$.data.failedFilenames[0]").value("invalid.txt"))
                .andExpect(jsonPath("$.message").value("다건 업로드 완료 (성공: 1건, 실패: 1건)"));
    }

    @Test
    @DisplayName("지원하지 않는 파일 형식이면 서비스 예외를 400으로 처리한다")
    void uploadImageInvalidExtension() throws Exception {
        MockMultipartFile txtFile = new MockMultipartFile(
                "file", "test.txt", "text/plain", "content".getBytes()
        );

        given(imageService.uploadImage(any(), anyLong()))
                .willThrow(new com.plog.global.exception.exceptions.ImageException(
                        com.plog.global.exception.errorCode.ImageErrorCode.INVALID_FILE_EXTENSION,
                        "지원하지 않는 파일 형식입니다."
                ));

        mockMvc.perform(multipart("/api/images").file(txtFile))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.status").value("fail"))
                .andExpect(jsonPath("$.message").value("지원하지 않는 파일 형식입니다."));
    }

    @Test
    @DisplayName("파일 없이 요청하면 400 Bad Request가 발생한다")
    void uploadImageWithoutFile() throws Exception {
        mockMvc.perform(multipart("/api/images"))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.message").value("Missing request part"));
    }
}

File: src/test/kotlin/com/plog/domain/image/controller/ImageControllerTest.kt
package com.plog.domain.image.controller

import com.ninjasquad.springmockk.MockkBean
import com.plog.domain.image.dto.ImageUploadRes
import com.plog.domain.image.service.ImageService
import com.plog.global.exception.errorCode.ImageErrorCode
import com.plog.global.exception.exceptions.ImageException
import com.plog.global.security.SecurityUser
import com.plog.testUtil.SecurityTestConfig
import com.plog.testUtil.WebMvcTestSupport
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.http.MediaType
import org.springframework.mock.web.MockMultipartFile
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import org.springframework.web.multipart.MultipartFile

@WebMvcTest(ImageController::class)
@ActiveProfiles("test")
@Import(SecurityTestConfig::class)
class ImageControllerTest : WebMvcTestSupport() {

    @MockkBean
    private lateinit var imageService: ImageService

    @BeforeEach
    fun setUp() {
        val mockUser = mockk<SecurityUser>(relaxed = true)
        every { mockUser.id } returns 1L
        SecurityContextHolder.getContext().authentication =
            UsernamePasswordAuthenticationToken(mockUser, null, emptyList())

        every { imageService.uploadImage(any<MultipartFile>(), any<Long>()) } returns
                ImageUploadRes(listOf("http://minio/uuid.jpg"), emptyList())
        every { imageService.uploadImages(any<List<MultipartFile>>(), any<Long>()) } returns
                ImageUploadRes(listOf("http://minio/bulk1.jpg", "http://minio/bulk2.jpg"), emptyList())
    }

    @Test
    @DisplayName("이미지 업로드 성공 시 successUrls를 포함한 응답을 반환한다")
    fun uploadImageSuccess() {
        val file = MockMultipartFile("file", "test.jpg", "image/jpeg", "test data".toByteArray())

        mockMvc.perform(multipart("/api/images").file(file).contentType(MediaType.MULTIPART_FORM_DATA))
            .andDo(print())
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.status").value("success"))
            .andExpect(jsonPath("$.data.successUrls[0]").value("http://minio/uuid.jpg"))
            .andExpect(jsonPath("$.message").value("이미지 업로드 성공"))
    }

    @Test
    @DisplayName("다중 이미지 업로드 성공 시 successUrls 리스트를 반환한다")
    fun uploadImagesSuccess() {
        val file1 = MockMultipartFile("files", "test1.jpg", "image/jpeg", "data1".toByteArray())
        val file2 = MockMultipartFile("files", "test2.jpg", "image/jpeg", "data2".toByteArray())

        mockMvc.perform(
            multipart("/api/images/bulk").file(file1).file(file2).contentType(MediaType.MULTIPART_FORM_DATA)
        ).andDo(print())
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.data.successUrls").isArray)
            .andExpect(jsonPath("$.data.successUrls.length()").value(2))
            .andExpect(jsonPath("$.message").value("다중 이미지 업로드 성공"))
    }

    @Test
    @DisplayName("다중 이미지 업로드 부분 실패 시 실패 파일명도 포함하여 반환한다")
    fun uploadImagesPartialFailure() {
        every { imageService.uploadImages(any<List<MultipartFile>>(), any()) } returns
                ImageUploadRes(listOf("http://minio/uuid1.jpg"), listOf("invalid.txt"))

        val file1 = MockMultipartFile("files", "test1.jpg", "image/jpeg", "data1".toByteArray())

        mockMvc.perform(multipart("/api/images/bulk").file(file1).contentType(MediaType.MULTIPART_FORM_DATA))
            .andDo(print())
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.data.failedFilenames[0]").value("invalid.txt"))
            .andExpect(jsonPath("$.message").value("다건 업로드 완료 (성공: 1건, 실패: 1건)"))
    }

    @Test
    @DisplayName("지원하지 않는 파일 형식이면 서비스 예외를 400으로 처리한다")
    fun uploadImageInvalidExtension() {
        every { imageService.uploadImage(any(), any()) } throws
                ImageException(ImageErrorCode.INVALID_FILE_EXTENSION, "지원하지 않는 파일 형식입니다.")

        val txtFile = MockMultipartFile("file", "test.txt", "text/plain", "content".toByteArray())
        mockMvc.perform(multipart("/api/images").file(txtFile))
            .andExpect(status().isBadRequest())
            .andExpect(jsonPath("$.status").value("fail"))
            .andExpect(jsonPath("$.message").value("지원하지 않는 파일 형식입니다."))
    }

    @Test
    @DisplayName("파일 없이 요청하면 400 Bad Request가 발생한다")
    fun uploadImageWithoutFile() {
        mockMvc.perform(multipart("/api/images"))
            .andExpect(status().isBadRequest())
            .andExpect(jsonPath("$.message").value("Missing request part"))
    }
}
